### ISSUE TITLE
 Visual Builder :: VB doesn't loading with WP Google Maps plugin

### ISSUE THREAD

**Issue by dm-chechik** [elegantthemes/Divi-Beta#453]
### Problem:

Visual builder doesn't loading with WP Google Maps shortcode on a page. I can insert shortcode into the page and see that maps appears but after page reload VB doesn't loading anymore.

### Steps To Reproduce:
1. Create new page and insert WP Google Maps shortcode
2. Save the page and reload it.
3. VB doesn't loading anymore on this page.

**Comment by fikrirasyid** [comment]
Hi @dm-chechik, I've investigated this issue. Basically, there are two things that causes the issue:

1. WP Google Maps doesn't anticipate how Divi VB works 
2. WP Google Maps preemptively removes all scripts that contains Google Maps API script to prevent Google Maps scripts duplication. Divi enqueues Google Maps script for its Map module. Divi VB uses this Google Maps scripts as one of its dependency. Because Divi's Google Maps script is removed, Divi VB scripts cannot be loaded. Hence the issue.

I've looked for ways to fix this issue. However, the cleanest way possible to fix this issue doesn't involve any modification in Divi. There are two things to do:

## 1. Modify WP Google Maps file 
This is needed to avoid null being treated as object which throws lots of error warning if you have `WP_DEBUG` enabled:

<img width="1331" alt="screen shot 2016-09-14 at 11 57 21 am" src="https://cloud.githubusercontent.com/assets/916442/18500676/e9dcd006-7a73-11e6-947e-bf2ed7dec87f.png">

Look at `/plugins/wp-google-maps/wpGoogleMaps.php`. Look at line `1640` then edit this:

<img width="580" alt="screen shot 2016-09-14 at 11 57 02 am" src="https://cloud.githubusercontent.com/assets/916442/18500661/cce09280-7a73-11e6-95c3-be9145891aca.png">

Into this:

<img width="528" alt="screen shot 2016-09-14 at 11 58 21 am" src="https://cloud.githubusercontent.com/assets/916442/18500664/d1d2b1f6-7a73-11e6-9755-8c650e47eb96.png">

## 2. Install this plugin then activate it

[wp-google-maps-divi-vb-compat.zip](https://github.com/elegantthemes/Divi-Beta/files/471317/wp-google-maps-divi-vb-compat.zip)

This is the code of the plugins:

```
function wpgmza_divivb_compat_load() {
	// Simplified version of is_fb_enabled(). Re-created due to
	// inability to reuse FB functions outside FB files
	if ( empty( $_GET['et_fb'] ) ) {
		return false;
	}

	if ( is_customize_preview() ) {
		return false;
	}

	if ( ! et_pb_is_allowed( 'use_visual_builder' ) ) {
		return false;
	}

	// Removed functions which removes maps.google enqueueing process
	remove_action('wp_enqueue_scripts', 'wpgmza_deregister_scripts',999);
	remove_action('wp_head', 'wpgmza_deregister_scripts',999);
	remove_action('init', 'wpgmza_deregister_scripts',999);
	remove_action('wp_footer', 'wpgmza_deregister_scripts',999);
	remove_action('wp_print_scripts', 'wpgmza_deregister_scripts',999);
}
add_action( 'init', 'wpgmza_divivb_compat_load' );
```

What this plugin does is a really simple:

1. If this page is opened in visual builder ...
2. Remove WP Google Maps' preemptive google maps script duplication which removes Divi's google maps script dependency

***
 
That's pretty much about it. 

Important notice: in VB, for the time being, shortcode is being rendered but it is not functional:

<img width="1370" alt="screen shot 2016-09-14 at 12 12 02 pm" src="https://cloud.githubusercontent.com/assets/916442/18500743/7a20528c-7a74-11e6-8cc6-93b0ff4761c6.png">

**Comment by fikrirasyid** [comment]
I'm closing this issue since the fix doesn't involves Divi and I think it is way too specific to be included into Divi source code.

@timayhelen If you think we need to include the fix into Divi source code, please let me know.

**Comment by dm-chechik** [comment]
Is it difficult to include some setting to Divi that disables Google Maps functionality? In this way we will enhance the compatibility with 3rd party plugins that provides more advanced functionality of maps. Also we have several clietns from China where Google Maps API blocked by the Great Firewall and VB is not loading.

**Comment by fikrirasyid** [comment]
@dm-chechik the issue isn't about difficult or easy, but:

## 1
it is the issue of "is this plugin widely used enough so we should add it to Divi" or not. If we're not considering the widely used factor and simply add all compat tweaks to Divi, we will end up add every compat issue to Divi and it isn't good for everyone.

## 2
Plus, the issue is actually caused by the way plugin assume that it is safe to modify other plugin/theme's enqueued scripts and replace it with its own file (which isn't enqueued using WordPress' enqueueing system and appended using javascript instead). What it doesn't assume is if other plugin/themes use the enqueued file as dependency of other script or not.

In short, this is what WP Google Maps do:

1. Checking "is there other plugins/themes which enqueues google maps' API scripts"?
2. Dequeueing it if it found any
3. Appended Google Maps API via javascript

IMO, what WP Google Maps should do, is:

1. Checking "is there other plugins/themes which enqueues google maps' API scripts"?
2. Check whether the enqueued Google Maps API script is used as dependency by other script. Important: if a an enqueued script is removed, all of its dependant wouldn't be load. Basically like removing the bottom part of house of cards. Thus why VB isn't loaded.
3. Dequeue found Google Maps API script if it isn't used as dependency of other plugin/theme. Don't dequeue if it is used as dependency. It is better to have duplicate loaded file and the WP configuration still works rather than optimized file load but breaking other plugin/themes mechanism.

***

I hope this helps explaining the issue :)

**Comment by lots0logs** [comment]
@fikrirasyid I've broadened the scope of this issue. Its no longer about solving a compatability issue with a single plugin. Based on the increasing number of tickets from members in China who cannot access google services (so they need a way to not have the gmaps script included on their pages), I think we can address that problem as well as the original problem of this issue by adding an option to disable google maps completely (eg. do not add the script to the page, and set the maps modules as disabled in role editor as a side effect of enabling the new option)

**Comment by fikrirasyid** [comment]
@lots0logs an option to enable / disable Google Maps option due to China customer makes sense, but I don't think it'll solve the WP Google Maps issue due to how WP Google Maps adds its Google Maps API scripts: It appends the file using Javascript, not enqueuing the script using `wp_enqueue_script`. So when VB access the `google` variable, it isn't loaded yet. 

Even if WP Google Maps API loads its Google Map API scripts using `wp_enqueue_script`, a plugin specific tweaks in inevitable (if we still want Divi Maps to work, unless disabling Google Maps API script === automatically disabling Maps module; this kinda makes sense tho) because we need to set it as VB's dependency.

We can still give it a try tho, what I explained above is based on what I know fixing this ticket's issue these past several days.

**Comment by lots0logs** [comment]
@fikrirasyid TBH I hadnt looked into the plugin issue at all. I was planning to follow your lead on that. Maybe we should open a separate issue for the new option to disable gmaps then?

>(if we still want Divi Maps to work, unless disabling Google Maps API script === automatically disabling Maps module; this kinda makes sense tho

I was thinking that enabling the option which disables gmaps would implicitly disable maps modules in the role editor.

**Comment by fikrirasyid** [comment]
@lots0logs yeah, I think it'd be better to open different issues for disable Google Maps  mechanism.

> I was thinking that enabling the option which disables gmaps would implicitly disable maps modules in the role editor.

I think It is logical to do so.

**Comment by fikrirasyid** [comment]
@lots0logs I'm opening https://github.com/elegantthemes/Divi-Beta/issues/532 for Google Map related issue and closing this one
